src/ssl/PeekingPeerConnector.cc

```diff
@@ -215,6 +218,10 @@ Ssl::PeekingPeerConnector::noteNegotiationDone(ErrorState *error)
     if (!request->clientConnectionManager.valid() || !fd_table[serverConnection()->fd].ssl)
         return;
 
+    // Abort if client connection is closed
+    if (!Comm::IsConnOpen(clientConn))
+        return;
```

Comment:
This check seems premature. Most of the code below does not use clientConn and may update transaction state. I suggest that we check clientConn only when we actually need/use it. In this method, that happens only when calling switchToTunnel(). The latter assumes an open connection -- we should add a Must() or assert() to switchToTunnel().

I am also concerned about the "do nothing and hope that the caller will figure it out" strategy. The comment says "abort" but we are aborting nothing. We simply do not do what the caller may expect us to do. Perhaps we should throw instead of returning?

Reply:
> This check seems premature. Most of the code below does not use clientConn and may successfully update transaction state with useful (e.g., logged server datails) information. I suggest that we check clientConn only when we actually need/use it.

Yep.


>  In this method, that happens only when we decide to call switchToTunnel(). As you know, the latter assumes an open client connection -- we should add the corresponding Must() or assert() to switchToTunnel(). Let's move this check just above the switchToTunnel() call?

Yes, I added this "Must". Moreover the clientConn does not needed in switchToTunnel function it can retrieve it from request->clientConnectionManager->clientConnection. The same for PeekingPeerConnector.  But this PR is already fixes many-many thinks.

> I am also concerned about the implicit "do nothing and hope that the caller will figure it out" strategy. The source code comment says "abort", but we are aborting nothing. We simply do not do what the caller may expect us to do. We could call mustStop() before returning, but it feels like we should throw instead of successfully returning. What do you think?

You are right. My only problem is the DBG_IMPORTANT debug message about a BUG inside PeerConnector::swanSong which will be printed in this case. 
Is it ok to replace this DBG_IMPORANT with an other to level 2 or 3?

Reply:
> My only problem is the DBG_IMPORTANT debug message about a BUG inside PeerConnector::swanSong which will be printed in this case. Is it ok to replace this DBG_IMPORANT with an other to level 2 or 3?

We do not have a formalized policy about this (yet), but I am leaning towards the following: In new code: Must()s are similar to asserts -- the code should not use Must()s as a reaction to a known/anticipated problem (e.g., a client connection closure). If we follow this principle, then the `BUG` message in swanSong() should stay at `DBG_IMPORTANT` -- it indicates an unexpected problem (that may result in memory leaks or be otherwise mishandled).

Please investigate whether noteNegotiationDone() can call bail() and _then_ throw? The former would be proper handling of the anticipated problem and the latter would be a proper action for a method that cannot complete as promised.

Reply:
> Please investigate whether noteNegotiationDone() can call bail() and then throw

Yep this is looks that works well.  I was stuck thinking of using the "Must" clause.
Commits 8d2a60c5 and 456ad0ab.

Reply:
Thank you for addressing my concern.